core/storage/job/gae_models.py

```diff
@@ -143,3 +143,55 @@ class ContinuousComputationModel(base_models.BaseModel):
     # The time at which a halt signal was last sent to this batch job, in
     # milliseconds since the epoch.
     last_stopped_msec = ndb.FloatProperty(indexed=True)
+
+
+# Allowed transitions: processing --> completed --> archived
+# or processing --> failed.
+QUERY_STATUS_PROCESSING = "processing"
+QUERY_STATUS_COMPLETED = "completed"
+QUERY_STATUS_ARCHIVED = "archived"
+QUERY_STATUS_FAILED = "failed"
+
+
+# Keys for query_parameters dictionary.
+# Values corresponding to key is value specified by query submitter.
+# A query can have any combination of following parameters.
+# Valid parameters names.
+# pylint: disable=C0103
+QUERY_PARAMETERS = utils.create_enum(
+    "has_user_logged_in_since_n_days", "created_less_than_n_exploration",
+    "created_more_than_n_exploration", "edited_less_than_n_exploration",
+    "edited_more_than_n_exploration",
+    "created_or_edited_exploration_in_n_days")
+
+class QueryModel(base_models.BaseModel):
+    """Model for storing result of queries.
+
+    The id of each instance of this model is alphanumeric id of length 12
+    unique to each model instance.
+    """
+    # parameters for query given by submitter.
+    query_parameters = ndb.JsonProperty()
+    # list of all user_ids who satisfy all parameters given in above query.
+    user_ids = ndb.StringProperty(repeated=True)
+    # id of the user who submitted the query.
+    submitter_id = ndb.StringProperty(indexed=True)
+    # current status of the query.
+    query_status = ndb.StringProperty(
+        indexed=True,
+        choices=[
+            QUERY_STATUS_PROCESSING,
+            QUERY_STATUS_COMPLETED,
+            QUERY_STATUS_ARCHIVED,
+            QUERY_STATUS_FAILED
+        ])
+
+    @classmethod
+    def create(cls, submitter_id, query_parameters_domain):
```

Comment:
I lean against this pattern, since it differs sufficiently from existing model-creation conventions (in particular, with regards to passing a domain object to a create() method in the storage model and converting it to a dict there). Can we instead add a create_user_query_model() in a bulk_email_services.py or user_query_services.py file in the domain layer (basically, whatever services file is going to handle the bulk email query page)?